File: src/gui/qt/log.cpp
/***************************************************************************
 *   Copyright (C) 2006 by FThauer FHammer   *
 *   f.thauer@web.de   *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/
#include "log.h"

#include "mainwindowimpl.h"

using namespace std;

Log::Log(mainWindowImpl* w) : myW(w)
{
	myW->setLog(this);

	myConfig = new ConfigFile;
	if(myConfig->readConfigString("LogDir") != "" && QDir::QDir(QString::fromStdString(myConfig->readConfigString("LogDir"))).exists()) { 

#ifdef _WIN32
		myLogDir = new QDir(QString::fromStdString(myConfig->readConfigString("LogDir")));
#else
		myLogDir = new QDir(QString::fromStdString(myConfig->readConfigString("LogDir")));
#endif
		myLogFile = new QFile(myLogDir->absolutePath()+"/pokerth-log-"+QDateTime::currentDateTime().toString("yyyy-MM-dd_hh.mm.ss")+".html");

		//Logo-Pixmap extrahieren
		QPixmap::QPixmap(":graphics/graphics/logo-140-100.png").save(myLogDir->absolutePath()+"/logo.png");

// 		myW->textBrowser_Log->append(myLogFile->fileName());

		myLogFile->open( QIODevice::WriteOnly );
		QTextStream stream( myLogFile );
		stream << "<html>\n";
		stream << "<body>\n";
		stream << "<img src='logo.png'>\n";
		stream << "<h3><b>Log-File for PokerTH Session started on "+QDate::currentDate().toString("yyyy-MM-dd")+" at "+QTime::currentTime().toString("hh:mm:ss")+"</b></h3>\n";
// 		stream << "</body>\n";
// 		stream << "</html>\n";
		myLogFile->close();

		linesInFile = 3;
	} 

	//Zu alte Dateien löschen!!!
	int daysUntilWaste = myConfig->readConfigInt("LogStoreDuration");
	int i;
		
	QStringList filters("pokerth-log*");
	QStringList logFileList = myLogDir->entryList(filters, QDir::Files);
	
	for(i=0; i<logFileList.count(); i++) {

// 		cout << logFileList.at(i).toStdString() << endl;

		QString dateString = logFileList.at(i);
		dateString.remove("pokerth-log-");
		dateString.remove(10,14);
		
		QDate dateOfFile(QDate::fromString(dateString, Qt::ISODate));
		QDate today(QDate::currentDate());
		
// 		cout << dateOfFile.daysTo(today) << endl;

		if (dateOfFile.daysTo(today) > daysUntilWaste) {

// 			cout << QString::QString(myLogDir->absolutePath()+"/"+logFileList.at(i)).toStdString() << endl;
			QFile fileToDelete(myLogDir->absolutePath()+"/"+logFileList.at(i));
			fileToDelete.remove();
		}

	}
	
}

Log::~Log()
{
}

void Log::logPlayerActionMsg(string playerName, int action, int setValue) {

	int i;	

	QString msg;
	msg = QString::fromStdString(playerName);
	
	switch (action) {

		case 1: msg += " folds.";
		break;
		case 2: msg += " checks.";
		break;
		case 3: msg += " calls ";
		break;
		case 4: msg += " bets ";
		break;
		case 5: msg += " sets ";
		break;
		case 6: msg += " is all in with ";
		break;
		default: msg += "ERROR";
	}
	
	if (action >= 3) { msg += QString::number(setValue,10)+"$."; }
	
	myW->textBrowser_Log->append(msg);

	myLogFile->open( QIODevice::ReadWrite );
	QTextStream stream( myLogFile );
	for(i=0; i<=linesInFile; i++) { stream.readLine(); }
	stream << msg+"</br>\n";
// 	stream << "</body>\n";
// 	stream << "</html>\n";
	myLogFile->close();

	linesInFile++;

}

void Log::logNewGameHandMsg(int gameID, int handID) {

	int i;

	myW->textBrowser_Log->append("<b>## Game: "+QString::number(gameID,10)+" | Hand: "+QString::number(handID,10)+" ##</b>");
	
	myLogFile->open( QIODevice::ReadWrite );
	QTextStream stream( myLogFile );
	for(i=0; i<=linesInFile; i++) { stream.readLine(); }
	
	stream << "<p><b>##### Game: "+QString::number(gameID,10)+" | Hand: "+QString::number(handID,10)+" #####</b></br>" << "\n";
	stream << "CASH: ";
	for(i=0; i<myW->getActualHand()->getActualQuantityPlayers(); i++) {

		if(myW->getActualHand()->getPlayerArray()[i]->getMyButton() == 1) {
			if(i == myW->getActualHand()->getActualQuantityPlayers()-1) {
				stream << QString::fromStdString(myW->getActualHand()->getPlayerArray()[i]->getMyName())+" (Dealer): "+QString::number(myW->getActualHand()->getPlayerArray()[i]->getMyCash(),10)+"$";
			}
			else {
				stream << QString::fromStdString(myW->getActualHand()->getPlayerArray()[i]->getMyName())+" (Dealer): "+QString::number(myW->getActualHand()->getPlayerArray()[i]->getMyCash(),10)+"$, ";
			}		
		}
		else {
			if(i == myW->getActualHand()->getActualQuantityPlayers()-1) {
				stream << QString::fromStdString(myW->getActualHand()->getPlayerArray()[i]->getMyName())+": "+QString::number(myW->getActualHand()->getPlayerArray()[i]->getMyCash(),10)+"$";
			}
			else {
				stream << QString::fromStdString(myW->getActualHand()->getPlayerArray()[i]->getMyName())+": "+QString::number(myW->getActualHand()->getPlayerArray()[i]->getMyCash(),10)+"$, ";
			}
		}
	}
	stream << "</br>BLINDS: ";
	for(i=0; i<myW->getMaxQuantityPlayers(); i++) {
	
		int j,k = 0;
		//Aktive Spieler zählen
		int activePlayersCounter = 0;
		for (k=0; k<myW->getMaxQuantityPlayers(); k++) { 
			if (myW->getActualHand()->getPlayerArray()[k]->getMyAction() != 1 && myW->getActualHand()->getPlayerArray()[k]->getMyActiveStatus() == 1) activePlayersCounter++;
		}
		if(activePlayersCounter < 3) { j=1; }
		
// 		cout << activePlayersCounter << endl;
// 		cout << (i+myW->getActualHand()->getDealerPosition()+j)%5 << endl;

		switch (myW->getActualHand()->getPlayerArray()[(i+myW->getActualHand()->getDealerPosition()+j)%5]->getMyButton()) {
			case 2 : stream << QString::fromStdString(myW->getActualHand()->getPlayerArray()[(i+myW->getActualHand()->getDealerPosition()+j)%5]->getMyName())+" ("+QString::number(myW->getActualHand()->getPlayerArray()[(i+myW->getActualHand()->getDealerPosition()+j)%5]->getMySet(),10)+"$), ";
			break;
			case 3 : stream << QString::fromStdString(myW->getActualHand()->getPlayerArray()[(i+myW->getActualHand()->getDealerPosition()+j)%5]->getMyName())+" ("+QString::number(myW->getActualHand()->getPlayerArray()[(i+myW->getActualHand()->getDealerPosition()+j)%5]->getMySet(),10)+"$)";	
			break;
			default :;	
		}
	}
	stream << "</p>\n";

	myLogFile->close();

	linesInFile = linesInFile+3;

}

void Log::logPlayerWinsMsg(int playerID) {

	int i;

	myW->textBrowser_Log->append(QString::fromStdString(myW->getActualHand()->getPlayerArray()[playerID]->getMyName())+" wins!!! ");
	
	myLogFile->open( QIODevice::ReadWrite );
	QTextStream stream( myLogFile );
	for(i=0; i<=linesInFile; i++) { stream.readLine(); }
	
	QString msg("<p><i>"+QString::fromStdString(myW->getActualHand()->getPlayerArray()[playerID]->getMyName())+" wins!!!</i></p>\n");
	stream << msg;

	myLogFile->close();

	linesInFile++;
	linesInFile++;

}

File: console/src/lib/Log.cs
/***************************************************************************
 *   Copyright (C) 2008 by Lothar May                                      *
*    Inspired by log.cpp of the PokerTH client which is                    *
 *   Copyright (C) 2006 by FThauer FHammer                                 *
 *                                                                         *
 *   This file is part of pokerth_console.                                 *
 *   pokerth_console is free software: you can redistribute it and/or      *
 *   modify it under the terms of the GNU Affero General Public License    *
 *   as published by the Free Software Foundation, either version 3 of     *
 *   the License, or (at your option) any later version.                   *
 *                                                                         *
 *   pokerth_console is distributed in the hope that it will be useful,    *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the                                *
 *   GNU Affero General Public License along with pokerth_console.         *
 *   If not, see <http://www.gnu.org/licenses/>.                           *
 ***************************************************************************/

using System;
using System.Collections.Generic;
using System.Text;

namespace pokerth_lib
{
	public class Log
	{
		public static string CardToString(int card)
		{
			int cardValue = card % 13;
			int cardColor = card / 13;
			string cardString;

			switch (cardValue)
			{
				case 0:
					cardString = "2";
					break;
				case 1:
					cardString = "3";
					break;
				case 2:
					cardString = "4";
					break;
				case 3:
					cardString = "5";
					break;
				case 4:
					cardString = "6";
					break;
				case 5:
					cardString = "7";
					break;
				case 6:
					cardString = "8";
					break;
				case 7:
					cardString = "9";
					break;
				case 8:
					cardString = "T";
					break;
				case 9:
					cardString = "J";
					break;
				case 10:
					cardString = "Q";
					break;
				case 11:
					cardString = "K";
					break;
				case 12:
					cardString = "A";
					break;
				default:
					cardString = "Invalid Card ";
					break;
			}
			switch (cardColor)
			{
				case 0 :
					cardString += "d";
					break;
				case 1 :
					cardString += "h";
					break;
				case 2 :
					cardString += "s";
					break;
				case 3:
					cardString += "c";
					break;
				default:
					cardString += " Invalid Color";
					break;
			}
			return cardString;
		}

		public static string StateToString(Hand.State state)
		{
			string stateString;

			switch (state)
			{
				case Hand.State.Preflop :
					stateString = "Preflop";
					break;
				case Hand.State.Flop :
					stateString = "Flop";
					break;
				case Hand.State.Turn :
					stateString = "Turn";
					break;
				case Hand.State.River :
					stateString = "River";
					break;
				default :
					stateString = "Invalid state";
					break;
			}

			return stateString;
		}

		public static string ActionToString(Hand.Action action, uint curSet)
		{
			string actionString;

			switch (action)
			{
				case Hand.Action.Fold:
					actionString = "folds";
					break;
				case Hand.Action.Check:
					actionString = "checks";
					break;
				case Hand.Action.Call:
					actionString = "calls";
					break;
				case Hand.Action.Bet:
					actionString = "bets";
					break;
				case Hand.Action.Raise:
					actionString = "bets";
					break;
				case Hand.Action.AllIn:
					actionString = "is all in with";
					break;
				default :
					actionString = "(invalid action)";
					break;
			}

			if ((int)action >= 3)
				actionString += " $" + Convert.ToString(curSet);

			return actionString;
		}

		protected static string CardCodeToString(int cardCode, bool plural)
		{
			string cardString;

			switch (cardCode)
			{
				case 12 :
					cardString = "Ace";
					break;
				case 11 :
					cardString = "King";
					break;
				case 10 :
					cardString = "Queen";
					break;
				case 9 :
					cardString = "Jack";
					break;
				case 8 :
					cardString = "Ten";
					break;
				case 7 :
					cardString = "Nine";
					break;
				case 6 :
					cardString = "Eight";
					break;
				case 5 :
					cardString = "Seven";
					break;
				case 4 :
					cardString = "Six";
					break;
				case 3 :
					cardString = "Five";
					break;
				case 2 :
					cardString = "Four";
					break;
				case 1 :
					cardString = "Three";
					break;
				case 0 :
					cardString = "Deuce";
				break;
				default :
					cardString = "Invalid Card";
					break;
			}
			if (plural)
			{
				if (cardCode == 4) // sixes
					cardString += "e";
				cardString += "s";
			}
			return cardString;
		}

		public static string CardsValueToString(int cardsValue)
		{
			string cardString;

			int firstPart = cardsValue / 100000000;
			int secondPart = cardsValue / 1000000 - firstPart * 100;
			int thirdPart = cardsValue / 10000 - firstPart * 10000
				- secondPart * 100;
			int fourthPart = cardsValue / 100 - firstPart * 1000000
				- secondPart * 10000 - thirdPart * 100;
			int fifthPart = cardsValue - firstPart * 100000000
				- secondPart * 1000000 - thirdPart * 10000 - fourthPart * 100;
			int fifthPartA = cardsValue / 10 - firstPart * 10000000
				- secondPart * 100000 - thirdPart * 1000 - fourthPart * 10;
			int fifthPartB = cardsValue - firstPart * 100000000
				- secondPart * 1000000 - thirdPart * 10000 - fourthPart*100
				- fifthPartA*10;

			switch (firstPart)
			{
				// Royal Flush
				case 9 :
					cardString = "Royal Flush";
					break;
				// Straight Flush
				case 8 :
					cardString = "Straight Flush, ";
					if (secondPart >= 3)
						cardString += CardCodeToString(secondPart, false) + " high";
					else
						cardString += "Invalid high";
					break;
				// Four of a Kind
				case 7 :
					cardString = "Four of a Kind, " + CardCodeToString(secondPart, true);
					break;
				// Full House
				case 6 :
					cardString = "Full House, "
						+ CardCodeToString(secondPart, true)
						+ " full of "
						+ CardCodeToString(thirdPart, true);
					break;
				// Flush
				case 5 :
					cardString = "Flush, ";
					if (secondPart >= 4)
						cardString += CardCodeToString(secondPart, false) + " high";
					else
						cardString += "Invalid high";
					break;
				// Straight
				case 4 :
					cardString = "Straight, ";
					if (secondPart >= 3)
						cardString += CardCodeToString(secondPart, false) + " high";
					else
						cardString += "Invalid high";
					break;
				// Three of a Kind
				case 3 :
					cardString = "Three of a Kind, " + CardCodeToString(secondPart, true);
					break;
				// Two Pairs
				case 2 :
					cardString = "Two Pairs, "
						+ CardCodeToString(secondPart, true)
						+ " and "
						+ CardCodeToString(thirdPart, true);
					break;
				// One Pair
				case 1 :
					cardString = "One Pair, "
						+ CardCodeToString(secondPart, true);
				break;
				// Highest Card
				case 0 :
					cardString = "High Card, "
						+ CardCodeToString(secondPart, false);
					break;
				default:
					cardString = "Invalid Card Value";
					break;
			}
			return cardString;
		}
	}
}
